gtda/curves/_functions.py

```diff
@@ -0,0 +1,52 @@
+# License: GNU AGPLv3
+
+import warnings
+from itertools import product
+
+import numpy as np
+from joblib import Parallel, delayed
+
+_AVAILABLE_FUNCTIONS = {
+    "identity": {},
+    "argmax": {},
+    "argmin": {},
+    "min": {},
+    "max": {},
+    "mean": {},
+    "std": {},
+    "median": {},
+    "average": {"weights": {"type": np.ndarray}}
+    }
+
+_implemented_function_recipes = {
+    "identity": lambda X, axis: X.reshape(len(X), -1),
+    "argmax": np.argmax,
+    "argmin": np.argmin,
+    "min": np.min,
+    "max": np.max,
+    "mean": np.mean,
+    "std": np.std,
+    "median": np.median,
+    "average": np.average
+    }
+
+
+def _parallel_featurization(Xt, function, function_params, n_jobs):
+    if callable(function):
+        return function(Xt, axis=-1, **function_params)
+    else:  # Assume function is a list or tuple of functions or None
+        channel_idx = [i for i, f in enumerate(function) if f is not None]
+        n_samples = len(Xt)
+        index_tups = product(*map(range, Xt.shape[:-2]), channel_idx)
```

Comment:
I miss the reason for the complexity here with the map and the tuple formulations. Would that not work to just have lists and do:

```
        Xt = Parallel(n_jobs=n_jobs)(
            delayed(function[c])(Xt[s, c], **function_params[c])
            for c in Xt.shape[1] for s in Xt.shape[0]
            )
```
and then some concatenate like we do usually?

Reply:
Ah yes you're right to note that. I wrote a version that worked on nd arrays just in case the discussion on PersistenceLandscape was still on the table. I agree though that if we settle on this not taking 4D arrays or higher, we should simplify the logic.

Reply:
I am not sure I follow. Even if the curve as 4D, all that should be done is in the beginning of `transform` to have:

```
if Xt.ndim == 4:
        Xt = Xt.reshape((Xt.shape[0], -1, Xt.shape[-1]))
```

And then it would be treated as all multi channel curves.

Reply:
I don't think the reshape you suggest is equivalent to what I had in mind.  For instance I don't think it would return what we would want in the case of `PersistenceLandscape`, because calculating the maxes of each layers and then flattening the result is not the same as concatenating the layers and then computing the maxes for each thus-flattened homology dimension.

When writing those lines I had in mind an even greater level of generality in which one passes an n-d array and gets out an (n-1)-d array or a 2D array depending on whether a `flatten` parameter is passed as `True` or not. But we are converging on it being out of scope for this transformer, so I am not seriously defending any of this!

Reply:
`curves.StandardFeatures` has to take in curves and return features. If there is a concrete interest in having a functional transformer that `sklearn.preprocessing.FunctionTransformer` does not cover, then, I suggest we could have a `curves.StandardFunctionTransformer`.

But TBH, I really wonder what anyone would do with a 3D arrays of persistent landscape layers maxes, apart from actually reshaping it to use them as features :). On transfromers, we should give options to users that makes sense and for which we provide support in the form of a family of next transformers that will handle meaningfully the data structure outputted.

Reply:
I agree with you that the extra generality is neither necessary nor appropriate here.  I just hadn't got round to committing the changes to avoid the `combinations` gymnastics.  I'm doing it now :)

Reply:
Ouffff :) Thanks! xD

Reply:
The logic is now much simpler and tied to the input being always 3d.